---
title: "Data resources"
output: github_document
---

```{r setup0, echo=FALSE, eval=TRUE, message=FALSE, warning=FALSE}
library(here)
library(yaml)
library(readxl)
library(tidyverse)
library(data.table)
library(vcfR)
library(gt)
library(RColorBrewer)
library(googlesheets4)
source("scripts/plot_utils.R")
config <- read_yaml("config.yaml")
cohort_labbeller <- c("familial"="Familial Breast","tcga"="TCGA-BRCA", "magic"="MAGIC","qimprove"="Q-IMPROvE")
extra_tumours <- config$extra_tumours
data_dir <- config$GRCh38_data_folder
knitr::opts_chunk$set(root.dir = here(), comment = NA, echo = FALSE, message = FALSE, warnings = FALSE, include = TRUE, fig.align = 'center', fig.retina = 2, fig.path = "figures/", cache.path ="cache/")
```

In this analysis, we included whole-genome sequencing (WGS) data of pairs of primary tumour and normal blood (germline) samples from breast cancer patients in four study cohorts. After filtering samples with low quality and sequencing depth, we had 350 tumours from:

- [Familial breast cancer](https://doi.org/10.1093/annonc/mdz132) (N=77)

- [TCGA-BRCA](https://portal.gdc.cancer.gov/projects/TCGA-BRCA) (N=96)

- [MAGIC](https://pubmed.ncbi.nlm.nih.gov/37005005/) (N=136)

- [Q-IMPROvE](https://pubmed.ncbi.nlm.nih.gov/36958289/) (N=41)

These datasets were used to call somatic variants from HRD prediction and to call germline variants for germline BRCA1 and BRCA2 classification. 

All samples were processed with a uniform pipeline and we applied the dual calling pipeline for analysing WGS data of tumour samples with matched germline samples to identify somatic SNV and indel changes. Specifically, we aligned the short-read sequencing reads to the human genome assembly (GRCh38) using BWA, and high confident somatic SNV were identified as the intersect of post-filtered output of qSNP (cite) and GATK HaplotypeCaller (cite). Short insertion and deletions (1-50bp) were called with the GATK HaplotyperCaller. Mutations were annotated with qannotate and SnpEff to identify somatic-specific mutations and filter mutations located within 6 base pairs of a homopolymer region.


```{r, message=FALSE}
# load grafli report
df_none2019 <- read_csv(paste0(data_dir,"Simpson_-_FamilialBreast_SomaticReport_GRCh38_77.csv")) |> filter(!is.na(ascatAnalysis)) |> 
  dplyr::rename("sample_id"=donorPublicationID)

uuid_tsv <- read_tsv(paste0(data_dir, "uuid_case.txt"),col_names = c("uuid","sample_id"))
df_tcga <- read_csv(paste0(data_dir, "TCGA_-_breast_cancer_SomaticReport_GRCh38.csv")) |> filter(!is.na(ascatAnalysis)) |> 
  left_join(uuid_tsv, by=c("donorSubmittedWithID"="uuid")) 

df_magic <- read_csv(paste0(data_dir, "Waddell_-_Magic_WGS_StudySomaticReport.csv.139")) |> filter(!is.na(ascatAnalysis)) |> dplyr::rename("sample_id"=donorSubmittedWithID, "tumour_id"=testSampleLabel) |> filter(!tumour_id %in% extra_tumours)

##  A single tumour had two germline controls as they were testing different kits, remove "collectedsample:4a6e39fd-4e92-496d-a1ce-98b1a9340d8c"
df_qimprove <- read_tsv(paste0(data_dir, "GenomiQa_-_qimprove_Somatic_Report_GRCh38_for_jia_no_recurrance.tsv")) |> 
  filter(! controlSampleIRI=="collectedsample:4a6e39fd-4e92-496d-a1ce-98b1a9340d8c") |> add_column(studyLabel="qimprove") |> dplyr::rename("sample_id"=patientLabel, "tumour_id"=testSampleLabel) |> 
  filter(!tumour_id %in% extra_tumours)
```

After removing samples with less then 20x depth in tumour sample, we only keep the primary tumour from each patient.

```{r sequencing_depth, fig.width=9, fig.height=4}
depth_df <- rbind(read_tsv(paste0(data_dir, "sequencing_depth/nones.depth.txt"), col_names = c("sample","t_depth","n_depth")) |> add_column(cohort="familial"),
                  read_tsv(paste0(data_dir, "sequencing_depth/tcga.depth.txt"), col_names = c("sample","t_depth","n_depth")) |> add_column(cohort="tcga"),
                  read_tsv(paste0(data_dir, "sequencing_depth/magic.depth.txt"), col_names = c("sample","t_depth","n_depth")) |> add_column(cohort="magic"),
                  read_tsv(paste0(data_dir, "sequencing_depth/qimprove.depth.txt"),col_names = c("sample","t_depth","n_depth")) |> add_column(cohort="qimprove")
                  ) |> 
  separate_wider_delim(sample, delim = ".", names = c("sample_id","tumour_id"),too_few = "align_start") |> 
  filter(t_depth>=20, !tumour_id %in% extra_tumours) |> select(-tumour_id)

sample_depth_order <- depth_df |> arrange(desc(t_depth)) |> pull(sample_id)

depth_df |> mutate(n_depth=-n_depth, sample_id=factor(sample_id,levels=sample_depth_order), cohort=factor(cohort,levels=c("familial","tcga","magic","qimprove"))) |>
  pivot_longer(c(t_depth,n_depth)) |> 
  ggplot(aes(x=sample_id,y=value,fill=name)) + 
  geom_col(alpha=.5) + 
  facet_grid(~cohort, scale="free_x", space = "free_x", labeller = labeller(cohort=cohort_labbeller)) + 
  geom_hline(yintercept = c(30,-30)) + 
  scale_y_continuous(breaks = c(-50,0,50,100), labels=c("50x","0","50x","100x"))+
  theme_Publication()+
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) +
  labs(x="",y="Sequencing depth",fill="") + 
  scale_fill_manual(values=c("#F0AB00", "#2A6EBB"), labels=c("Normal", "Tumour"))

#low_depth_samples <- depth_df |> filter(t_depth<15) |> pull(sample)

valid_samples <- depth_df |> pull(sample_id)
```

A figure showing the tumour purity scores, TMB in cohorts.

```{r supplementary_fig1ab, fig.width=9.6, fig.height=4.5, warning=FALSE, message=FALSE}
df_all <- rbind(df_none2019 |> select(studyLabel,sample_id,ascatPurityScore,TMB_nonSyn),
      df_tcga |> select(studyLabel,sample_id, ascatPurityScore,TMB_nonSyn), 
      df_magic |> select(studyLabel,sample_id,ascatPurityScore,TMB_nonSyn),
      df_qimprove |> select(studyLabel, sample_id, ascatPurityScore,TMB_nonSyn) ) |> 
  mutate(cohort=case_when(studyLabel=="Simpson_-_FamilialBreast" ~ "familial",
                          studyLabel=="TCGA_-_breast_cancer" ~ "tcga",
                          studyLabel=="Waddell_-_Magic_WGS" ~ "magic",
                          TRUE ~ studyLabel),
        cohort=factor(cohort,levels=c("familial","tcga","magic","qimprove"))) |> 
  mutate(sample_id = str_split_i(sample_id,"\\.",i=1)) |> filter(sample_id %in% valid_samples)

#df_all |> select(sample_id, ascatPurityScore, TMB_nonSyn)  |> write_tsv("results/purity_tmb.tsv")

df_all |> pivot_longer(c(ascatPurityScore,TMB_nonSyn)) |> 
  mutate(name=factor(name,levels=c("ascatPurityScore","TMB_nonSyn"),labels=c("Tumour purity","Tumour mutation burden (TMB)"))) |> 
  ggplot(aes(x=cohort,y=value)) + 
  geom_violin() + geom_boxplot(width=0.1) +
  scale_x_discrete(labels=cohort_labbeller) +
  facet_wrap(~name, scales = "free_y") + 
  labs(x="",y="Scores") + theme_Publication()

df_all |> 
  group_by(cohort) |> 
  summarise(mean_purity = mean(ascatPurityScore), min=min(ascatPurityScore), max=max(ascatPurityScore)) |> gt() |> as_raw_html()
```
 
 
 
A figure showing the number of snvs and indels in each cohort.

```{r num_of_variants, fig.width=9.2, fig.height=4.6, message=FALSE, warning=FALSE, results='hide', cache=TRUE}
summarise_vcfs <- function(pdir, sample) {
  snv_file <- file.path("data/GRCh38",pdir,"snvs",paste0(sample,".snv.vcf.gz"))
  indel_file <- file.path("data/GRCh38",pdir,"indels",paste0(sample,".indel.vcf.gz"))
  n_snv <- read.vcfR(snv_file) |> nrow()
  n_indel <- read.vcfR(indel_file) |> nrow()
  data.frame(snv=n_snv,indel=n_indel)
}


df_vcf_all <- rbind(
  df_none2019$sample_id |> set_names() |> map(\(x) summarise_vcfs("Nones2019",x)) |> list_rbind(names_to = "sample_id"),
  df_tcga |> filter(sample_id %in% valid_samples) |> pull(sample_id) |> set_names() |> map(\(x) summarise_vcfs("TCGA_BRCA",x)) |> list_rbind(names_to = "sample_id"),
  df_magic |> filter(sample_id %in% valid_samples, !tumour_id %in% extra_tumours)  |> unite(sample_id, c("sample_id","tumour_id"),sep = ".") |> pull(sample_id) |> set_names(\(x) str_split_i(x,"\\.",i=1)) |> map(\(x) summarise_vcfs("MAGIC",x)) |> list_rbind(names_to = "sample_id"),
  df_qimprove |> filter(sample_id %in% valid_samples, !tumour_id %in% extra_tumours)  |> unite(sample_id, c("sample_id","tumour_id"),sep = ".") |> pull(sample_id) |> set_names(\(x) str_split_i(x,"\\.",i=1)) |> map(\(x) summarise_vcfs("Qimprove",x)) |> list_rbind(names_to = "sample_id") 
  ) 

df_vcf_all |> mutate(cohort=case_when(grepl("FBC",sample_id) ~ "familial",
                                                       grepl("TCGA",sample_id) ~ "tcga",
                                                       grepl("MAGIC",sample_id) ~ "magic",
                                                       grepl("GQ",sample_id) ~ "qimprove"), 
                                      cohort=factor(cohort,levels=c("familial","tcga","magic","qimprove")),
                                      sampl_id=factor(sample_id,levels=sample_depth_order)) |> 
  pivot_longer(c(snv,indel)) |> 
  mutate(name=factor(name,levels=c("snv","indel"))) |> 
  ggplot(aes(x=sample_id,y=value/1000)) + geom_col() + 
  facet_grid(name~cohort, scales = "free", space="free_x", labeller = labeller(cohort=cohort_labbeller, name=c("snv"="SNV", "indel"="Indel"))) + 
  scale_y_continuous(labels = \(x) ifelse(x==0,x,paste0(x,"k"))) + 
  theme_Publication() + labs(x="Sample",y="Count") + 
  theme(axis.ticks.x = element_blank(), axis.text.x = element_blank()) 
```

```{r supplementary_fig1c, fig.width=9.2, fig.height=3.6, warning=FALSE}
sv_df <- list.files("data/GRCh38", pattern = "*.sv.tsv", full.names = TRUE, recursive = TRUE) |> 
  set_names(\(x) basename(x) |> str_remove(".sv.tsv")) |> map(\(x) read_tsv(x,col_types = list(col_character(),col_double()))) |> 
  list_rbind(names_to = "sample") |> mutate(sv_len=ifelse(is.na(sv_len), 0, sv_len)) |> 
  mutate(cohort=case_when(grepl("FBC",sample) ~ "familial",
                          grepl("TCGA",sample) ~ "tcga",
                          grepl("MAGIC",sample) ~ "magic",
                          grepl("GQ",sample) ~"qimprove"),
         cohort=factor(cohort,levels=c("familial","tcga","magic","qimprove"))) |> 
  separate_wider_delim(sample, delim = ".", names = c("sample_id","tumour_id"),too_few = "align_start") |> 
  filter(!tumour_id %in% extra_tumours) |> 
  mutate(sample_id=factor(sample_id,levels=sample_depth_order))

sv_df |> ggplot(aes(x=sample_id,fill=sv_type)) + geom_bar(stats="count") + 
  facet_grid(~cohort, scales = "free_x",space = "free_x", labeller = labeller(cohort=cohort_labbeller)) + ylim(0,1000) +
  scale_fill_manual(values = c("#386cb0","#f87f01","#7fc97f","#ef3b2c"), labels=c("DEL","INS/DUP","INV","TRA")) +
  theme_Publication() + labs(x="Sample",y="SV count", fill="") + 
  theme(axis.ticks.x = element_blank(), axis.text.x = element_blank(), legend.position = "bottom")
```

All these data has been presented in [Supplementary table 1](https://docs.google.com/spreadsheets/d/1rgva1pfGp3ScCln8a0WScAAEoazS1-dUu5NfeF_1AD0/edit?usp=sharing).

```{r, write_google_sheet}
google_sheet_supp1 <- sv_df |> group_by(sample_id) |> summarise(total_sv=n()) |> ungroup() |> 
  full_join(df_vcf_all, by="sample_id") |> 
  full_join(df_all) |> 
  full_join(depth_df) |> 
  replace_na(list(total_sv=0)) |> 
  mutate(Study=case_when(studyLabel=="Simpson_-_FamilialBreast" ~ "Familial Breast",
                          studyLabel=="TCGA_-_breast_cancer" ~ "TCGA-BRCA",
                          studyLabel=="Waddell_-_Magic_WGS" ~ "MAGIC",
                          studyLabel=="qimprove" ~ "Q-IMPROvE")) |> 
  select(sample_id, Study, t_depth, n_depth, ascatPurityScore, TMB_nonSyn, snv, indel, total_sv) |> 
  arrange(Study) |> 
  mutate(sample_id=ifelse(startsWith(sample_id,"MAGIC"), magic_id_conversion[sample_id], sample_id)) |> 
  dplyr::rename("Sample ID" = sample_id,"SV" = total_sv, "Tumour depth" = t_depth, "Normal depth" = n_depth, "Tumour purity" = ascatPurityScore, "TMB" = TMB_nonSyn, "SNV" = snv, "Indel" = indel, "SV" = total_sv) 

## add a column to indicate which sample has RNA seq data.
tpm_df <- read_tsv("../x.revision_misc/brca12_tpm.tsv")
google_sheet_supp1 <- google_sheet_supp1 |> left_join(tpm_df, by=c("Sample ID" = "sample_id")) |> mutate(RNAseq=ifelse(is.na(BRCA1_TPM), "No", "Yes")) |> select(-c(BRCA1_TPM, BRCA2_TPM))


google_sheet_supp1 |> head(10) |> 
  gt() |> as_raw_html()
#google_sheet_supp1 |> write_tsv("results/supplementary-table1.rev.tsv")
```

```{r, eval=FALSE}
#don't want to set up this as non interative auth at the moment.
library(httr)
set_config(use_proxy(url=config$proxy_url,port=8080, username = config$proxy_user, password = config$proxy_passwd))

gs4_auth()

google_sheet_supp1 |> 
  write_sheet(
    ss = gs4_get(
      config$google_sheet_id
    ),
    sheet = "Supplementary Table 1"
  )
```


