---
title: "Substitution mutational signatures"
output: github_document
---

```{r setup0, echo=FALSE, eval=TRUE, message=FALSE, warning=FALSE}
library(here)
library(tidyverse)
library(deconstructSigs)
library(BSgenome.Hsapiens.UCSC.hg19)
```

### de novo mutational signature extraction using NMF

```{r, warning=FALSE}
read_maf <- function(maf_file) {
  sample_id <- basename(maf_file) %>% str_remove(".maf.gz")
  read_tsv(maf_file, col_types = cols(.default = "c"), comment="#") %>% add_column(sample = sample_id) 
}

maf_df <- list.files("data/Nones/mafs","maf.gz$",full.names = T) %>% map_dfr(read_maf)

maf_df %>% filter(Variant_Type=="SNP") %>% count(sample)
```

```{r}
snv_df<- maf_df %>% filter(Variant_Type=="SNP", !grepl("GL|Y|MT",Chromosome)) %>%
  mutate(Sample=sample,chr=paste0("chr",Chromosome),pos=as.integer(Start_Position),ref=Reference_Allele, alt=ifelse(Tumor_Seq_Allele1==Reference_Allele, Tumor_Seq_Allele2, Tumor_Seq_Allele1)) %>%
  select((last_col()-5+1):last_col()) %>% as.data.frame()

sigs.input <- mut.to.sigs.input(mut.ref = snv_df,
                                bsg = BSgenome.Hsapiens.UCSC.hg19)

cosmic2_t<-read_tsv("data/cosmic/COSMIC_v2_SBS_GRCh37.txt")  %>%
  arrange(match(Type,colnames(sigs.input))) %>% 
  column_to_rownames("Type") %>% t 

cosmic_sigs <- as.data.frame(cosmic2_t) %>% 
  rownames_to_column(var = "Signature") %>%
  mutate(Signature = str_replace(Signature,"Signature_","Sig")) %>%
  column_to_rownames(var = "Signature")

rownames(cosmic2_sigs) <- gsub("Signature_","Sig",x = rownames(cosmic2_sigs))
```

```{r}
sigs_all <- data.frame()
for (s in rownames(sigs.input)){
 sigs <- whichSignatures(tumor.ref = sigs.input, signatures.ref = cosmic_sigs, sample.id = s,contexts.needed = TRUE, tri.counts.method = 'default',signature.cutoff = 0.1)
 sigs_all <- rbind(sigs_all, sigs$weight)
}

sigs_detected <- sigs_all %>%
    rownames_to_column(var="Donor")%>%
    gather(Sig, Count, -Donor) %>%
    group_by(Sig) %>%
    filter(Count > 0) %>%
    spread(Sig, Count, fill = 0)


df <- sigs_detected %>% pivot_longer(-Donor) %>% left_join(brca_status)
df$BRCA_status <- factor(df$BRCA_status, levels= c("non-BRCA","BRCA2","BRCA1"))
#df$name <- factor(df$name, levels = rev(c("Sig8","Sig3","Sig18","Sig1-like","Sig13","Sig16","Sig2","Sig5")))
  ggplot(df, aes(x=Donor, y= value, fill=name))+
  geom_bar(stat="identity",position = "fill") + 
  facet_wrap(~BRCA_status, scales = "free_x") + 
  scale_fill_d3() +
  theme_bw(base_size = 12) + labs(x="",y="",fill="") +
  theme(axis.text.x = element_blank(),axis.ticks.x = element_blank(), 
        legend.position = "bottom",legend.margin =margin(0,0,0,0))
  
 # ggsave("figures/fig-cosmic2-fit.png",width = 8.6, height = 3.4,dpi = 300)
```





