---
title: "HRDetect HRD prediction"
output: github_document
---

```{r setup0, echo=FALSE, eval=TRUE, message=FALSE, warning=FALSE}
library(here)
library(yaml)
library(tidyverse)
library(signature.tools.lib)
library(ggh4x)
source("scripts/signature.tools.lib-utils.R")
source("scripts/plot_utils.R")
config <- read_yaml("config.yaml")
extra_tumours <- config$extra_tumours
knitr::opts_chunk$set(root.dir = here(), comment = NA, echo = FALSE, message = FALSE, warnings = FALSE, include = TRUE, fig.align = 'center', fig.retina = 2, fig.path = "figures/", cache.path = "cache/")
```

```{r}
cosmicv2_signatures  <- read_table("data/COSMIC/COSMIC_v2_SBS_GRCh38.txt") |> column_to_rownames("Type")
cosmicv2_signatures <- cosmicv2_signatures[match(rownames(COSMIC30_subs_signatures), rownames(cosmicv2_signatures)),]
```

```{r, eval=FALSE}
sigsToUse <- c(1,2,3,5,6,8,13,17,18,20,26,30)
# for imap() the first param is the element of list, and the second is the index or name of the element.
sbs_df <- list.files("data/GRCh38", pattern = "*.snv.vcf.gz$",full.names = TRUE,recursive = TRUE) |>
  set_names(\(x) basename(x) |> str_remove(".snv.vcf.gz")) |> 
  imap(~Zainal_SBS_fit_vcf(sample = .y, snv_file=.x, signature_to_fit = cosmicv2_signatures[sigsToUse], hg="hg38")) |> 
  list_rbind(names_to = "sample")

sbs_df |> as.data.frame() |> write_tsv("results/signature_tools_lib_sbs.355.tsv")
col_hrdetect <- c("del.mh.prop", "SNV3", "SV3", "SV5", "hrd", "SNV8")

input_matrix_v2 <- matrix(data=NA, nrow = nrow(sbs_df), ncol = length(col_hrdetect), dimnames = list(sbs_df$sample, col_hrdetect))
input_matrix_v2[,"SNV3"] <- sbs_df$Signature_3
input_matrix_v2[,"SNV8"] <- sbs_df$Signature_8

indel_files <- list.files("data/GRCh38",pattern = "*.indel.vcf.gz$", full.names = TRUE, recursive = TRUE) |> (\(x) set_names(x, str_remove(basename(x), ".indel.vcf.gz") ))() 
cnv_files <- list.files("data/GRCh38", pattern = "*.cnv.txt$", full.names = TRUE, recursive = TRUE) |> (\(x) set_names(x, str_remove(basename(x), ".cnv.txt") ))() 
sv_files <- list.files("data/GRCh38",pattern = "*.sv.bedpe$", full.names = TRUE, recursive = TRUE) |> (\(x) set_names(x, str_remove(basename(x), ".sv.bedpe") ))() 


res_hg38 <- HRDetect_pipeline(input_matrix_v2,
                            genome.v = "hg38",
                            SV_bedpe_files = sv_files,
                            Indels_vcf_files = indel_files,
                            CNV_tab_files = cnv_files,
                            organ = "Breast",
                            nbootFit = 1000,
                            nparallel = 4)

write_tsv(res_hg38$hrdetect_output |> as.data.frame() |> rownames_to_column("sample"), "results/hrdetect_hrdprediction_355.tsv")
```


```{r hrdetect_fig1, fig.width=4, fig.height=4}
df_hrd <- read_tsv("results/hrdetect_hrdprediction_355.tsv") |> 
  select(sample, Probability) |> 
  separate_wider_delim(sample, ".", names = c("sample_id","tumour_id"), too_few = "align_start") |> 
  filter(!tumour_id %in% extra_tumours) |> 
  select(sample_id, Probability) |> 
  arrange(desc(Probability)) |> 
  mutate(sample_id=factor(sample_id, levels=sample_id)) 
p_hrdetect_threshold <- df_hrd |> 
  ggplot(aes(x=sample_id, y=Probability )) + geom_point(alpha=.6, color="#3B4992") + 
  geom_hline(yintercept = 0.7, color="grey", lty=2, linewidth=1) +  annotate("text", x=260, y=0.78, label="cutoff=0.7", size=3.6, color="black") +
  labs(x="Tumor samples", y= stringr::str_wrap("HRD probability (HRDetect)", 15)) +
  theme_Publication() +
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) 

p_hrdetect_threshold
```

**Figure 1: The HRD probability scores for all tumor samples**

```{r hrdetect_fig2, fig.height=4, fig.width=9}
df_hrd |> 
  mutate(cohort=case_when(grepl("FBC",sample_id) ~ "Familial Breast",
                                     grepl("TCGA",sample_id) ~ "TCGA-BRCA",
                                     grepl("MAGIC",sample_id) ~ "MAGIC",
                                     TRUE ~ "Q-IMPROvE"),
         cohort=factor(cohort, levels=cohort_levels))  |> 
  ggplot(aes(x=sample_id, y=Probability )) + geom_point(alpha=.6, color="#3B4992") + 
  geom_hline(yintercept = 0.7, color="grey", lty=2, linewidth=1) + facet_grid(~cohort, scales = "free_x",space = "free_x") + 
  theme_Publication() +
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) + labs(x="Tumor samples", y="HRD probability (HRDetect)")
```

**Figure 2: The HRD probability scores for tumour samples in four cohorts**


```{r hrdetect_fig3, fig.width=7, fig.height=4.5, warning=FALSE}
df_hrd |> 
  mutate(cohort=case_when(grepl("FBC",sample_id) ~ "Familial Breast",
                                     grepl("TCGA",sample_id) ~ "TCGA-BRCA",
                                     grepl("MAGIC",sample_id) ~ "MAGIC",
                                     TRUE ~ "Q-IMPROvE"),
         cohort=factor(cohort, levels=cohort_levels)) |> 
  mutate(type=ifelse(Probability>=0.7, "HRD","HRP")) |> 
  ggplot(aes(x=type,fill=type)) + geom_bar() + 
  geom_text(stat='count', aes(label=..count..), vjust=-.1) +
  facet_grid(~cohort) + ylim(0,130) +
  theme_Publication() + labs(x="", y= "Number of samples", fill="HR status") + scale_fill_manual(values=c("#0073C2","#EFC000")) +
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())
```

**Figure 3: The number of individuals identified as HRD or non-HRD in four cohorts by HRDetect**


Since MAGIC cohort was composed of FFPE-derived samples, we did [FFPE correction](FFPE_correction.md) to exam whether this affect our HRD prediction.


```{r, eval=FALSE}
ffpe_corr_sbs_df <- read_tsv("results/magic_ffpe_corrected_sbs.txt")

## re-run HRD pipeline for corrected MAGIC data.
input_matrix <- matrix(data=NA, nrow = nrow(ffpe_corr_sbs_df), ncol = length(col_hrdetect), dimnames = list(ffpe_corr_sbs_df$sample, col_hrdetect))
input_matrix[,"SNV3"] <- ffpe_corr_sbs_df$Signature_3
input_matrix[,"SNV8"] <- ffpe_corr_sbs_df$Signature_8

indel_files <- list.files("data/GRCh38/MAGIC",pattern = "*.indel.vcf.gz$", full.names = TRUE, recursive = TRUE) |> set_names(\(x) basename(x) |> str_split_i("\\.",i=2) )
cnv_files <- list.files("data/GRCh38/MAGIC", pattern = "*.cnv.txt$", full.names = TRUE, recursive = TRUE) |> set_names(\(x) basename(x) |> str_split_i("\\.",i=2) )
sv_files <- list.files("data/GRCh38/MAGIC",pattern = "*.sv.bedpe$", full.names = TRUE, recursive = TRUE) |> set_names(\(x) basename(x) |> str_split_i("\\.",i=2) )


res_magic_ffpe <- HRDetect_pipeline(input_matrix,
                            genome.v = "hg38",
                            SV_bedpe_files = sv_files,
                            Indels_vcf_files = indel_files,
                            CNV_tab_files = cnv_files,
                            organ = "Breast",
                            nbootFit = 1000,
                            nparallel = 4)

#res_magic_ffpe$hrdetect_output |> as.data.frame() |> rownames_to_column("tumour_id") |> write_tsv("results/magic_ffpe_corrected_hrdetect.txt")
```

Comparing the HRD prediction results before and after FFPE signature correction, there is no impact to number of HRD samples except some minor value changes.

```{r hrdetect_fig4, fig.height=4, fig.width=5}
df_hrd_magic <- read_tsv("results/hrdetect_hrdprediction_355.tsv") |> 
  filter(startsWith(sample,"MAGIC")) |> 
  separate_wider_delim(sample,names = c("sample_id","tumour_id"), delim = ".") |> 
  select(sample_id,tumour_id, Probability)

df_hrd_magic_corrected <- 
  read_tsv("results/magic_ffpe_corrected_hrdetect.txt") |> 
  select(tumour_id,Probability)

df_hrd_magic |> left_join(df_hrd_magic_corrected, by="tumour_id") |> 
  rename("Original"=Probability.x,"FFPE-corrected"=Probability.y) |> 
  filter(!tumour_id %in% extra_tumours) |> 
  arrange(desc(Original)) |> 
  mutate(sample_id=factor(sample_id, levels=sample_id)) |> 
  pivot_longer(c(Original, `FFPE-corrected`)) |> 
  mutate(name=factor(name, levels=c("Original", "FFPE-corrected"))) |> 
  ggplot(aes(x=sample_id,y=value, color=name)) + geom_point(alpha=.4, size=1.6) +
  geom_hline(yintercept = 0.7, color="grey", lty=2, linewidth=1) +
  labs(x="Tumor samples",y="HRD Probability (HRDetect)", color="") + theme_Publication() +
  scale_color_manual(values=c("#3B4992", "#EE0000")) +
  theme(legend.position = "bottom", axis.text.x = element_blank(), axis.ticks.x=element_blank())
```

**Figure 4: The HRD prediction scores before and after FFPE signature correction**

