---
title: "HRD prediction using CHORD"
output: github_document
---

```{r setup0, echo=FALSE, eval=TRUE, message=FALSE, warning=FALSE}
library(tidyverse)
library(yaml)
library(CHORD)
library(BSgenome.Hsapiens.UCSC.hg38)
library(RColorBrewer)
library(here)
library(gt)
library(waffle)
library(ggmosaic)
library(colorspace)
source("scripts/plot_utils.R")
config <- read_yaml("config.yaml")
extra_tumours <- config$extra_tumours
knitr::opts_chunk$set(root.dir = here(), comment = NA, echo = FALSE, message = FALSE, warnings = FALSE, include = TRUE, fig.align = 'center', fig.retina = 2, fig.path = "figures/", cache.path ="cache/")
```


```{r}
snv_files <- list.files("data/GRCh38", pattern = "*.snv.vcf$",recursive = TRUE, full.names = TRUE) |> set_names(\(x) basename(x) |> str_remove(".snv.vcf"))
indel_files <- list.files("data/GRCh38", pattern = "*.indel.vcf$",recursive = TRUE, full.names = TRUE) |> set_names(\(x) basename(x) |> str_remove(".indel.vcf"))
sv_df_list <- list.files("data/GRCh38", ".sv.tsv", recursive = TRUE,full.names = TRUE) |>  set_names(\(x) basename(x) |> str_remove(".sv.tsv")) |>  map(data.table::fread)
```

```{r cache=TRUE, eval=FALSE}
df_chord <- names(snv_files)|> 
  map_df(\(x) extractSigsChord(vcf.snv=snv_files[x], vcf.indel=indel_files[x], df.sv = sv_df_list[[x]], sample.name = x, ref.genome = BSgenome.Hsapiens.UCSC.hg38) |> as.data.frame()) |> 
  chordPredict(do.bootstrap = T, verbose = F)

write_tsv(df_chord, "results/chord_hrdprediction_355.tsv")
```

```{r}
df_chord <- read_tsv("results/chord_hrdprediction_355.tsv")
df_hrd <- df_chord[1:6] |> mutate(cohort=case_when(
  startsWith(sample, "MAGIC") ~ "MAGIC",
  startsWith(sample, "FBC") ~ "Familial Breast",
  startsWith(sample, "TCGA") ~ "TCGA-BRCA",
  startsWith(sample, "GQ") ~ "Q-IMPROvE",
  TRUE ~ "Unknown"
),
  hr_status = ifelse(hr_status=="HR_deficient", "HRD", "HRP")) |> 
  separate_wider_delim(sample,".", names = c("sample_id","tumour_id"), too_few = "align_start") |> 
  filter(!tumour_id %in% extra_tumours) |> select(-tumour_id) |> 
  arrange(desc(p_hrd) ) |> mutate(sample_id=factor(sample_id,levels=sample_id),
                                  cohort=factor(cohort, levels=cohort_levels))
```

```{r chord_fig1, fig.width=4, fig.height=4}
p_chord_threshold<- ggplot(df_hrd, aes(x=sample_id, y=p_hrd)) + geom_point(alpha=.6, color="#3B4992") + 
  geom_hline(yintercept = 0.5, color="grey", lty=2, linewidth=1) + annotate("text", x=260, y=0.58, label="cutoff=0.5", size=3.6, color="black") +
  theme_Publication() + labs(x="Tumor samples",y=stringr::str_wrap("HRD probability (CHORD)", 15)) + 
  theme(legend.position = "bottom", axis.text.x = element_blank(), axis.ticks.x = element_blank())

p_chord_threshold
```

**Figure 1: The HRD probability scores for tumour samples**

```{r chord_fig2, fig.height=4, fig.width=9}
ggplot(df_hrd, aes(x=sample_id, y=p_hrd)) + geom_point(alpha=.6, color="#3B4992") + 
  facet_grid(~cohort, scales = "free_x", space = "free_x") +
  geom_hline(yintercept = 0.5, color="grey", lty=2, linewidth=1) +
  theme_Publication()  + labs(x="Tumor samples",y="HRD probability (CHORD)") + 
  theme(legend.position = "bottom", axis.text.x = element_blank(), axis.ticks.x = element_blank())
```

**Figure 2: The HRD probability scores for tumour samples in four cohorts**

```{r chord_fig3, fig.width=7, fig.height=4.5, warning=FALSE}
df_hrd |> 
  mutate(cohort=case_when(grepl("FBC",sample_id) ~ "Familial Breast",
                                     grepl("TCGA",sample_id) ~ "TCGA-BRCA",
                                     grepl("MAGIC",sample_id) ~ "MAGIC",
                                     TRUE ~ "Q-IMPROvE"),
         cohort=factor(cohort, levels=c("Familial Breast","TCGA-BRCA","MAGIC","Q-IMPROvE"))) |> 
  ggplot(aes(x=hr_status,fill=hr_status)) + geom_bar() + 
  geom_text(stat='count', aes(label=..count..), vjust=-.1) +
  facet_grid(~cohort) + ylim(0,130) +
  theme_Publication() + labs(x="", y= "Number of samples", fill="HR status") + scale_fill_manual(values=c("#0073C2","#EFC000")) +
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())
```

**Figure 3: The number of individuals identified as HRD or non-HRD in four cohorts by CHORD**

CHORD can distinguish BRCA1 and BRCA2 type in HRD group.

```{r figure3a, warning=FALSE, fig.width=4, fig.height=4.6}
my_blues <- brewer.pal(3,"Blues") |> rev()
my_blues[1:2] <- c("#005B96", "#00A8E1")
pa <- df_hrd |> filter(hr_status=="HRD") |> 
  ggplot(aes(x=cohort, fill=hrd_type)) + geom_bar() + scale_fill_manual(values=my_blues, labels=c("BRCA1","BRCA2","Undetermined")) +
  geom_text(data=df_hrd |> filter(hr_status=="HRD") |>  filter(cohort!="MAGIC" | hrd_type!="BRCA1_type", cohort!="Q-IMPROvE" | hrd_type=="BRCA1_type"), aes(label=..count..), stat = "count", vjust = 1.5, position = "stack", size=3) + 
  geom_text(data=df_hrd |> filter(hr_status=="HRD") |>  filter(cohort=="MAGIC" & hrd_type=="BRCA1_type"), aes(label=..count..), stat = "count", vjust = -4.1, position = "stack", size=3) + 
  geom_text(data=df_hrd |> filter(hr_status=="HRD") |>  filter(cohort=="Q-IMPROvE" & hrd_type=="BRCA2_type"), aes(label=..count..), stat = "count", vjust = -.6, position = "stack", size=3) + 
  geom_text(data=df_hrd |> filter(hr_status=="HRD") |>  filter(cohort=="Q-IMPROvE" & hrd_type=="cannot_be_determined"), aes(label=..count..), stat = "count", vjust = .6, position = "stack", size=3) + 
  theme_Publication() + labs(x="",y="Number of samples", fill="HRD type") + 
  theme(legend.position = "bottom", axis.text.x = element_text(size=8), legend.text = element_text(size=8), legend.title = element_text(size=9)) 
pa
```

## CHORD HRD prediction after FFPE impact correction
```{r, cache=TRUE}
snv_files <- list.files("data/GRCh38/MAGIC", pattern = "*.snv.vcf$",recursive = TRUE, full.names = TRUE) |> set_names(\(x) basename(x) |> str_remove(".snv.vcf"))
indel_files <- list.files("data/GRCh38/MAGIC", pattern = "*.indel.vcf$",recursive = TRUE, full.names = TRUE) |> set_names(\(x) basename(x) |> str_remove(".indel.vcf"))
sv_df_list <- list.files("data/GRCh38/MAGIC", ".sv.tsv", recursive = TRUE,full.names = TRUE) |>  set_names(\(x) basename(x) |> str_remove(".sv.tsv")) |>  map(data.table::fread)

magic_chord_context <- names(snv_files)|> map_df(\(x) extractSigsChord(vcf.snv=snv_files[x], vcf.indel=indel_files[x], df.sv = sv_df_list[[x]], sample.name = x, ref.genome = BSgenome.Hsapiens.UCSC.hg38) |> as.data.frame())

corrected_mutation <- read_tsv("data/MAGIC.ffpe_unrepaired_corrected.extractor_sorted.matrix.txt")

magic_chord_context[,1:96] <- corrected_mutation[match(colnames(magic_chord_context[,1:96]), corrected_mutation$MutationType),] |> 
  column_to_rownames("MutationType") |> t() |> as.data.frame()

chord_magic_ffpe <- chordPredict(magic_chord_context, do.bootstrap=T, verbose=F) |> select(sample, p_hrd) |> separate_wider_delim(sample,".", names = c("sample_id","tumour_id")) |> 
  filter(!tumour_id %in% extra_tumours) |> select(-tumour_id)

chord_magic <- df_chord |> filter(startsWith(sample,"MAGIC")) |> select(sample, p_hrd) |> separate_wider_delim(sample,".", names=c("sample_id","tumour_id")) |> 
  filter(!tumour_id %in% extra_tumours) |> select(-tumour_id)
```

```{r chord_fig5, fig.height=4, fig.width=5}
inner_join(chord_magic, chord_magic_ffpe, by="sample_id") |> 
  dplyr::rename("Original"=p_hrd.x,"FFPE-corrected"=p_hrd.y) |> 
  arrange(desc(Original)) |> 
  mutate(sample_id=factor(sample_id, levels=sample_id)) |> 
  pivot_longer(-sample_id) |> mutate(name=factor(name, levels=c("Original", "FFPE-corrected"))) |> 
  ggplot(aes(x=sample_id,y=value, color=name)) + 
  geom_point(alpha=.4, size=1.6) +
  geom_hline(yintercept = 0.5, color="grey", lty=2, linewidth=1) +
  labs(x="Tumor samples",y="HRD Probability (CHORD)", color="") + theme_Publication() + 
  scale_color_manual(values=c("#3B4992", "#EE0000")) +
  theme(legend.position = "bottom", axis.text.x = element_blank(), axis.ticks.x=element_blank())
```

**Figure 4: The HRD prediction scores before and after FFPE signature correction**

## Using CHORD results to aid germline variant classification

The likelihood ratio test was performed to evaluate how strong if having a germline BRCA1/2mutation increases the likelihood of having somatic HRD based on number of HRD samples in each germline group with VUS BRCA1/2 group and P/LP or VUS in other genes excluded.

```{r mosaic_plot, fig.height=4.6, fig.width=5, warning=FALSE}
result_df <- read_tsv("results/brca_germline_clf_tumor_hrd.add_sv.add_somatic.rev.cf6.tsv")
mosaic_df <-result_df |> select(cohort,sample_id, updated_germline_classification, hr_status, hrd_type) |> 
  filter(updated_germline_classification %in% c("positive_brca1","positive_brca2","negative_brca1/2")) |> 
  mutate(hr_status=ifelse(hr_status=="HR_deficient","HRD","HRP"),
         updated_germline_classification=factor(updated_germline_classification, levels=c("negative_brca1/2","positive_brca1","positive_brca2")))

ggplot(mosaic_df) + geom_mosaic(aes(x=product(hr_status,updated_germline_classification), fill=hr_status)) + 
  geom_mosaic_text(aes(x = product(hr_status, updated_germline_classification), label = after_stat(.wt)), as.label=TRUE, size=3.2) +
  scale_fill_manual(values=c("#0073C2","#EFC000")) + theme_Publication() +
  labs(x="", y="", fill="HR status") + 
   scale_y_productlist(
    labels = c("HRD","HRP"),
    expand = c(0, 0)
  ) +
   scale_x_productlist(
    labels = c("BRCA1/2 negative",stringr::str_wrap("BRCA1 positive", 6),stringr::str_wrap("BRCA2 positive", 6)),
    expand = c(0, 0.01)
  ) +
  theme(axis.text.x=element_text(size=8), 
        axis.line.y = element_blank(), 
        axis.line.x = element_blank(), 
        axis.text.y=element_text(size=7.8),
        axis.ticks = element_blank(),
        legend.text = element_text(size=8), 
        legend.title = element_text(size=9), legend.margin = margin(0,0,0,0), legend.box.margin=margin(-10,-10,-10,-10))
```

if we include HRD subtypes:

```{r figure3b, fig.height=4.6, fig.width=5, warning=FALSE}
pb <- mosaic_df |> 
  filter(hrd_type!="cannot_be_determined") |> 
  ggplot() + geom_mosaic(aes(x=product(hrd_type, updated_germline_classification), fill=hrd_type)) + 
  geom_mosaic_text(aes(x = product(hrd_type,updated_germline_classification), label = after_stat(.wt)), as.label=TRUE, size=3.2) +
  scale_fill_manual(values=c(my_blues[1:2], "#EFC000"), labels=c("HRD(BRCA1)","HRD(BRCA2)","HRP")) +
  labs(x="", y="", fill="HR status") + 
  scale_y_productlist(
    labels = c("HRD (BRCA1)","HRD (BRCA2)","HRP"),
    expand = c(0, 0)
  ) +
   scale_x_productlist(
    labels = c("BRCA1/2 negative",stringr::str_wrap("BRCA1 positive", 6),stringr::str_wrap("BRCA2 positive", 6)),
    expand = c(0, 0.01)
  ) +
  theme_Publication() + 
  theme(axis.text.x=element_text(size=8), 
        axis.line.y = element_blank(), 
        axis.line.x = element_blank(), 
        axis.text.y=element_blank(),
        axis.ticks = element_blank(),
        legend.text = element_text(size=8), 
        legend.title = element_text(size=9), legend.margin = margin(0,0,0,0), legend.box.margin=margin(-10,-10,-10,-10))

pb
#ggsave("figures/figure3b.pdf", pb, height = 4.6, width = 5)
#ggsave("figures/figure3b.rev.pdf", pb, height = 4.6, width = 5)
```


31 individuals had been excluded due to carrying P/LP in non-BRCA1/2 genes, out of these 29 individuals, 5 of them are HRD (BRCA1), 4 are HRD (BRCA2) and 2 of them are HRD (Undetermined).  9 BRCA1/2 VUS cases were identified as HRD, 7 in BRCA1 type and 2 in BRCA2 type.

```{r figure 3c, fig.height=5.2, fig.width=9}
group_color_mapping2 <- c("additional"="#0099B4", "exclude_plp" = "#E5735A","exclude" = "#FDAF91", "negative_brca1/2"="#ADB6B6", "positive_brca1" = "#00468B", "positive_brca2"="#ED0000")
group_labeller2 <- c("BRCA1/2 VUS","Exclude P/LP","Exclude VUS","BRCA1/2 negative", "BRCA1 positive", "BRCA2 positive")
names(group_labeller2) <- c("additional","exclude_plp","exclude","negative_brca1/2","positive_brca1","positive_brca2")
hrdtype_labeller <- c(stringr::str_wrap("HRD (BRCA1)",8),stringr::str_wrap("HRD (BRCA2)",8),stringr::str_wrap("HRD (Undetermined)",10), "HRP")
names(hrdtype_labeller) <- c("BRCA1_type","BRCA2_type","cannot_be_determined","none")

transform_variant <- function(variant) {
  v <- str_split(variant, ";", simplify = TRUE)
  v |> map_chr(\(x) str_replace(x,
    pattern = "c\\.([^\\[]*).*\\(([A-Z0-9]+)\\)",
    replacement = "\\2 (c.\\1)"
  ) ) |> paste(collapse = "; ")
}

exclude_plp_df <- result_df |> select(cohort, sample_id, hrd_type, updated_germline_classification, B, class_B_mutations, pathSV) |> filter(B>0, hrd_type!="none") |> 
  mutate(variant=ifelse(is.na(class_B_mutations), 
                        paste0(str_split_i(pathSV,"_",i=5)," (DEL)"),
                        class_B_mutations |> map_chr(transform_variant))
         ) |> select(hrd_type, variant) |>  group_by(hrd_type) |> mutate(row=row_number()) |> ungroup() |> 
  mutate(col=case_when(hrd_type=="BRCA1_type"~11, hrd_type=="BRCA2_type"~8,TRUE~2),col=ifelse(hrd_type=="BRCA2_type" & row==7, 13.6 , col), row=ifelse(hrd_type=="BRCA2_type" & row==7, 6, row), row=ifelse(hrd_type=="BRCA1_type" & row==5, 6, row))


additional_df <- 
  result_df |> select(cohort, sample_id, hrd_type, updated_germline_classification, C, class_C_mutations, pathSV) |> 
  filter(grepl("additional", updated_germline_classification), hrd_type!="none") |> 
  mutate(variant=class_C_mutations |> map_chr(transform_variant)) |> 
  select(hrd_type, variant) |> mutate(text_length=str_length(variant)) |> 
  group_by(hrd_type) |> arrange(desc(text_length)) |> 
  mutate(row=row_number()) |> ungroup() |> 
  mutate(col=case_when(row==7~32.6, hrd_type=="BRCA1_type"~25.5, hrd_type=="BRCA2_type"~17), row=ifelse(hrd_type=="BRCA1_type" & row!=1, row+1, row))

result_df |> select(cohort, sample_id, hrd_type, updated_germline_classification, B, class_B_mutations, class_C_mutations,C, pathSV) |> filter(B>0 | C>0) |> 
  filter(hrd_type!="none", startsWith(updated_germline_classification,"additional")) |> 
  mutate(variant=str_replace_all(class_C_mutations, "c\\.(.*)\\((BRCA\\d+)\\)", "\\2 (\\1)")) |> 
  select(hrd_type, variant) |> gt() |> as_raw_html()
  


exclude_plp_text_color <- darken("#E5735A", amount = 0.3) ## kind of match with the box color but darker:)
additional_text_color <- darken("#0099B4", amount=0.3)
pc<-result_df |> select(cohort,sample_id, hrd_type, updated_germline_classification, A, B) |> 
  mutate(germline_classification=ifelse(startsWith(updated_germline_classification,"additional"),"additional",updated_germline_classification)) |> 
  mutate(detail_exclude=ifelse(B>0, "exclude_plp", germline_classification)) |> 
  mutate(detail_exclude=factor(detail_exclude, levels=c("positive_brca1","positive_brca2","exclude_plp","exclude","additional","negative_brca1/2"))) |> 
  group_by(hrd_type, detail_exclude) |> summarise(n=n()) |> ungroup() |> 
  ggplot() + 
  geom_waffle(aes(fill=detail_exclude, values=n),color = "white", size = 1.125, n_rows = 6) + 
  geom_text(data=exclude_plp_df, aes(label=variant, x=col,y=row), hjust = 0,color=exclude_plp_text_color, fontface="bold", size=rel(2.5)) +
  geom_text(data=additional_df, aes(label=variant, x=col,y=row), hjust = 0,color=additional_text_color, fontface="bold", size=rel(2.5)) +
  scale_fill_manual(values=group_color_mapping2, labels=group_labeller2) + scale_x_discrete(expand=c(0.01,0.1)) +
  facet_wrap(~hrd_type, ncol = 1,  strip.position = "left", labeller = labeller(hrd_type=hrdtype_labeller)) + theme_Publication() + labs(x="", y="HR status ",fill="Individual classification") +
  theme( legend.position = "inside", 
        legend.position.inside  = c(0.88,0.46), 
         legend.direction = "vertical",
            legend.box = "vetical",
         axis.line.y= element_blank(), axis.ticks = element_blank() ,axis.line.x = element_blank(), axis.text = element_blank(), strip.text = element_text(size=8),legend.text = element_text(size=8), legend.title = element_text(size=9))
pc
#ggsave("figures/figure3c.pdf", pc, height = 5.2, width = 9, dpi = 320)
#ggsave("figures/figure3c.rev.pdf", pc, height = 5.2, width = 9, dpi = 320)
```



